compiler/parser/Parser.y

```diff
@@ -1157,8 +1157,38 @@ topdecl :: { LHsDecl GhcPs }
 -- Templates
 --
 template_decl :: { OrdList (LHsDecl GhcPs) }
-  : 'template' qtycon arecord_with 'where' template_body
+  : 'template' template_header arecord_with 'where' template_body
                                                  {% mkTemplateDecls $2 $3 $5 }
+  | 'template' 'instance' qtycon '=' qtycon qtycons
+                                                 {% mkTemplateInstance $3 $5 (unLoc $6) }
+
+template_header :: { Located TemplateHeader }
```

Comment:
See the NOTE below. I tried that in this commit https://github.com/digital-asset/ghc/pull/32/commits/5dc09d9d072f5be9d7b887e6b481994110aa7885, but it caused failure to parse non-generic templates. It could be recoverable though, with some rejigging of the parse rules perhaps. It would be nice to parse a context directly, instead of parsing the constraints and converting them to an LHsContext later.

Reply:
Let's merge it as is then but try to improve in a subsequent PR. I'm convinced it is possible to reuse GHC's rules for this.